pkg/serviceaccounttoken/secret.go

```diff
@@ -5,41 +5,61 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
 )
 
-const serviceAccountSecretAnnotation = "kubernetes.io/service-account.name"
+const (
+	// ServiceAccountSecretLabel is the label used to search for the secret belonging to a service account.
+	ServiceAccountSecretLabel = "cattle.io/service-account.name"
 
-// secretGetter is an abstraction over any kind of secret getter.
+	serviceAccountSecretAnnotation = "kubernetes.io/service-account.name"
+)
+
+// secretLister is an abstraction over any kind of secret lister.
 // The caller can use any cache or client it has available, whether that is from norman, wrangler, or client-go,
 // as long as it can wrap it in a simplified lambda with this signature.
-type secretGetter func(namespace, name string) (*v1.Secret, error)
+type secretLister func(namespace string, selector labels.Selector) ([]*v1.Secret, error)
 
 // EnsureSecretForServiceAccount gets or creates a service account token Secret for the provided Service Account.
 // For k8s <1.24, the secret is automatically generated for the service account. For >=1.24, we need to generate it explicitly.
-func EnsureSecretForServiceAccount(ctx context.Context, secretGetter secretGetter, clientSet kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
-	if secretGetter == nil {
-		secretGetter = func(namespace, name string) (*v1.Secret, error) {
-			return clientSet.CoreV1().Secrets(namespace).Get(ctx, name, metav1.GetOptions{})
-		}
-	}
+func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrollers.SecretCache, clientSet kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
 	if sa == nil {
 		return nil, fmt.Errorf("could not ensure secret for invalid service account")
 	}
 	secretClient := clientSet.CoreV1().Secrets(sa.Namespace)
-	saClient := clientSet.CoreV1().ServiceAccounts(sa.Namespace)
-	secretName := ServiceAccountSecretName(sa)
-	var secret *v1.Secret
-	var err error
-	if secretName != "" {
-		secret, err = secretGetter(sa.Namespace, secretName)
-		if err != nil && !apierror.IsNotFound(err) {
-			return nil, fmt.Errorf("error ensuring secret for service account [%s:%s]: %w", sa.Namespace, sa.Name, err)
+	var secretLister secretLister
+	if secretsCache != nil {
+		secretLister = secretsCache.List
+	} else {
+		secretLister = func(_ string, selector labels.Selector) ([]*v1.Secret, error) {
+			secretList, err := secretClient.List(ctx, metav1.ListOptions{
+				LabelSelector: selector.String(),
+			})
+			if err != nil {
+				return nil, err
+			}
+			result := make([]*v1.Secret, len(secretList.Items))
+			for i := range secretList.Items {
+				result[i] = &secretList.Items[i]
+			}
+			return result, nil
+		}
+	}
+	secret, err := ServiceAccountSecret(sa, secretLister)
+	if err != nil {
+		return nil, fmt.Errorf("error looking up secret for service account [%s:%s]: %w", sa.Namespace, sa.Name, err)
+	}
+	if secret != nil && !isSecretForServiceAccount(secret, sa) {
```

Comment:
Nit: It's a bit inefficient (though this function isn't doing much) to check if the value `isSecretForServiceAccount` twice (both here and below). There's probably a few ways that you could fix this, but I would recommend at minimum saving this value above so you don't need to call the function more than once.

Reply:
:+1: